src/Type.js

```diff
@@ -0,0 +1,193 @@
+ * the modified value rather than mutating the previous instance.
+ * ```js
+ * var x1y0 = origin.x(1);
+ * x1y0.x(); // 1
+ *
+ * // note `origin` remains unmodified
+ * origin.x(); // 0
+ * ```
+ *
+ * When a type is created, lenses for each field are also available on the
+ * type constructor:
+ * ```js
+ * R.view(Point.x, point); // 0
+ * R.set(Point.y, point);  // Point(0, 1);
+ * ```
+ *
+ * ## Sum Types
+ *
+ * Sum types can be created by calling `sum` with an object that describes
+ * its constructors.
+ *
+ * For example will create a `List` type, with two constructors `Cons` and `Nil`:
+ * ```js
+ * var List = Type.sum({ Cons: ['head', 'tail'], Nil: [] });
+ * var Cons = List.Cons;
+ * var Nil = List.Nil;
+ * ```
+ *
+ * The constructors behave exactly the same as product type constructors:
+ * ```js
+ * var abc = Cons('a', Cons('b', Cons('c', Nil)));
+ * abc instanceof Cons; // true
+ * abc instanceof List; // true
+ *
+ * var c = abc.tail().tail().head();
+ * // or
+ * var c = R.view(R.compose(Cons.tail, Cons.tail, Cons.head), abc);
+ *
+ * abc.unapply(function(head, tail) {
+ *   return head + head + head;
+ * }); // "aaa"
+ * ```
+ *
+ * If a constructor is declared with no fields, a singleton instance
+ * of that constructor is returned instead:
+ * ```js
+ * Nil instanceof List; // true
+ * Nil.unapply(function() { return '?'; }); // '?'
+ * ```
+ *
+ * An instance can also be matched and dispatched against its type:
+ * ```js
+ * function headOrElse(other, list) {
+ *   return list.match({
+ *     Cons: function(head, tail) { return head; },
+ *     Nil: function() { return other; }
+ *   });
+ * }
+ * ```
+ */
+
+// Allows a given `proto` to be used as the base prototype of the product
+// instances -- this is used by `sum` to extend the generated sum type without
+// resorting to setting __proto__ or relying on Object.setPrototypeOf introduced
+// in ES6.
+function productOf(proto) {
+  // The actual function used to create a product type with the given set of field names.
+  return function() {
+    var field, i;
+    var fields = arguments;
+    var Constructor = R.curryN(fields.length, function () {
+      var vals = arguments;
+      return Object.create(Constructor.prototype, {
+        unapply: {
+          value: function unapply(f) {
+            return f.apply(this, vals);
+          }
+        }
+      });
+    });
+    Constructor.prototype = Object.create(proto);
+    Constructor.prototype.constructor = Constructor;
+    for (i = 0; i < fields.length; i++) {
+      field = fields[i];
+      Constructor[field] = productLens(i);
+      Constructor.prototype[field] = getterSetter(field);
+    }
+    return Constructor;
+  };
+}
+
+// The function used to create a sum type with the given constructor descriptor.
+function sum(constructors) {
+  var Type = function () {};
+
+  function buildProto(name) {
+    return Object.create(Type.prototype, {
+      match: {
```

Comment:
I'm not entirely familiar with the nuances of catamorphisms, but this method is named `cata` in other similar implementations and it does make sense in terms of folding the `Type` down to some other value. Perhaps @joneshf might be able to provide some insight?

I did however use the opportunity to follow through [this intro to F-Algebras and catamorphisms](https://www.fpcomplete.com/user/bartosz/understanding-algebras), translating the examples to JS using `Type` from this PR: https://gist.github.com/scott-christopher/35568367b82db7ef22a1 :smile: